misc-scripts/xref_mapping/XrefParser/HPAParser.pm

```diff
@@ -45,71 +79,60 @@ sub run {
   my $verbose    = $ref_arg->{verbose} // 0;
   my $dbi        = $ref_arg->{dbi} // $self->dbi;
 
-  croak
-    "Need to pass source_id, species_id, files and rel_file as pairs"
-    unless defined $source_id and
-    defined $species_id and
-    defined $files;
+  if ( ( !defined $source_id ) or
+       ( !defined $species_id ) or
+       ( !defined $files ) ) {
+    confess 'Need to pass source_id, species_id, and files';
+  }
 
   my $file = @{$files}[0];
 
   my $file_io = $self->get_filehandle($file);
-  croak "Could not open $file\n" unless defined $file_io;
+  if ( !defined $file_io ) {
+    confess "Could not open $file\n";
+  }
 
-  my $input_file =
-    Text::CSV->new( { sep_char => ",", empty_is_undef => 1 } ) or
-    croak "Cannot use file $file: " . Text::CSV->error_diag();
+  my $input_file = Text::CSV->new({
+    sep_char       => q{,},
+    empty_is_undef => 1,
+    strict         => 1,
+  }) or confess "Cannot use file $file: " . Text::CSV->error_diag();
 
-  $file_io->getline()
-    ;    # skip header # Antibody,antibody_id,ensembl_peptide_id,link
+  my @expected_columns = qw( Antibody antibody_id ensembl_peptide_id link );
+  my $header           = $input_file->getline($file_io);
 
+  if ( scalar @{$header} != scalar @expected_columns ) {
+    confess "input file $file has an incorrect number of columns";
+  }
+
+  $input_file->column_names(@expected_columns);
```

Comment:
This seems dangerous, shouldn't the column names be set from `@{$header}`? That way, if the format changes there is at least a chance for the parser to get the right data if the change is a simple rearrangement.

Reply:
Guess it was Mark's Python background (in Python attempting to read a dict value corresponding to a nonexistent key results in an exception) showing :-)

Anyway, I have opted for a more strict header validation procedure as implemented in the other parsers I have worked on. Moreover, I have got rid of _getline_hr()_ because it is known to be slow.